# Pré-processamento final dos dados
> O intuito desse notbook é sumarizar o processo para gerar os nossos dataframes, e os processamentos finais para criar o `bares_e_restaurantes.csv` e o `cdb.csv`, que tem o dados de todos os bares + os dados específicos dos bares participantes do Comida di Buteco.

In [1]:
! pip install geopandas
! pip install fiona==1.9.6
! pip install dash dash-leaflet geopandas pandas

# ! pip install --upgrade geopandas
# ! pip show geopandas
# ! pip show fiona

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
  Using cached fiona-1.9.6.tar.gz (411 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [3 lines of output]
      <string>:86: DeprecationWarning: The 'warn' function is deprecated, use 'warning' instead
      CRITICAL:root:A GDAL API version must be specified. Provide a path to gdal-config using a GDAL_CONFIG environment variable or use a GDAL_VERSION environment variable.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a sub

In [ ]:
# imports
import re
import pandas as pd
from pyproj import Transformer

## Fazendo o dataframe dos bares

In [ ]:
# definindo as colunas - coluna coords_geo

df =  pd.read_csv("dados_com_coordenadas.csv")

# define o transformador de UTM Zone 23S (EPSG:31983) para WGS 84 (EPSG:4326)
transformer = Transformer.from_crs("EPSG:31983", "EPSG:4326", always_xy=True)

def converter_utm_para_latlon(geom_str):
    match = re.search(r"POINT \((-?[\d\.]+) (-?[\d\.]+)\)", geom_str)
    if match:
        x = float(match.group(1))
        y = float(match.group(2))
        lon, lat = transformer.transform(x, y)
        return (lat, lon)
    return None  

df["COORD_GEO"] = df["GEOMETRIA"].apply(converter_utm_para_latlon)

# print(df.dtypes)
# df.head()

In [ ]:

for col in [
    "ID_ATIV_ECON_ESTABELECIMENTO",
    "CNAE_PRINCIPAL",
    "DATA_INICIO_ATIVIDADE",
    "GEOMETRIA"
]:
    df[col] = df[col].replace("indisponivel", "NA")

df["NOME_FANTASIA"] = df.apply(
    lambda row: row["NOME"] if row["NOME_FANTASIA"] == "indisponivel" else row["NOME_FANTASIA"],
    axis=1
)

df["ID_ATIV_ECON_ESTABELECIMENTO"] = df["ID_ATIV_ECON_ESTABELECIMENTO"].astype(str)
df["ID_CDB"] = df["CDB"].astype(int).apply(lambda x: f"{x:03}")
df.drop(columns=["CDB"], inplace=True)
df["CNAE_PRINCIPAL"] = df["CNAE_PRINCIPAL"].astype(str)

df["DATA_INICIO_ATIVIDADE"] = df["DATA_INICIO_ATIVIDADE"].apply(
    lambda x: pd.to_datetime(x, format="%d-%m-%Y").date() if x != "NA" else "NA"
)


for col in [
    "IND_POSSUI_ALVARA",
    "ENDERECO_COMPLETO",
    "NOME",
    "NOME_FANTASIA",
    "GEOMETRIA"
]:
    df[col] = df[col].astype(str)

def parse_coord_str(x):
    if isinstance(x, str):
        coords = tuple(map(float, x.strip().replace("(", "").replace(")", "").split(",")))
    elif isinstance(x, (tuple, list)):
        coords = x
    elif isinstance(x, float) or isinstance(x, int):
        coords = (float(x), float(x))
    else:
        coords = (float('nan'), float('nan'))
    
    return tuple(round(c, 7) for c in coords)

df["COORDS"] = df["COORDS"].apply(parse_coord_str)
df["COORD_GEO"] = df["COORD_GEO"].apply(parse_coord_str)



cols = df.columns.tolist()
cols.insert(1, cols.pop(cols.index("ID_CDB")))
df = df[cols]

# print(df.dtypes)
# df.head()

ID_ATIV_ECON_ESTABELECIMENTO    object
ID_CDB                          object
CNAE_PRINCIPAL                  object
DATA_INICIO_ATIVIDADE           object
IND_POSSUI_ALVARA               object
ENDERECO_COMPLETO               object
NOME                            object
NOME_FANTASIA                   object
GEOMETRIA                       object
COORDS                          object
COORD_GEO                       object
dtype: object


In [73]:
df.to_csv("complete_bar_data.csv", index=False, encoding="utf-8")

## Fazendo o dataframe dos dados do "Comida di Buteco"

In [78]:
df1 = pd.read_csv("../comida_di_buteco/cdb_data.csv")
df2 =  pd.read_csv("../comida_di_buteco/bares_cdb2025.csv")

In [ ]:
df2["Nome"] = df2["Nome"].str.strip('"')

df1["Nome_upper"] = df1["Nome"].str.upper()
df2["Nome_upper"] = df2["Nome"].str.upper()

df1["CDB_ID"] = df1["CDB_ID"].apply(lambda x: f"{int(x):03d}")

df_merged = pd.merge(df1, df2, on="Nome_upper", how="left")

df_merged = df_merged.drop(columns=["Nome_upper"])
df_merged.rename(columns={"Nome": "NOME"}, inplace=True)

df_merged["NOME"] = df_merged["Nome_x"]
df_merged["ENDERECO"] = df_merged["Endereço_x"]


df_merged.rename(columns={
    "Imagem": "IMAGEM",
    "Petisco": "PETISCO",
    "Descrição do petisco": "DESCRICAO"
}, inplace=True)

df_final = df_merged[["CDB_ID", "NOME", "ENDERECO", "IMAGEM", "PETISCO", "DESCRICAO"]]

df_final.to_csv("complete_cdb_data.csv", index=False, encoding="utf-8")

df_final.head()


CDB_ID                 NOME  \
0    001            ARCOS BAR   
1    002   AZOUGUE FOGO E BAR   
2    003               BAIÚCA   
3    004  BAR ESTABELECIMENTO   
4    005    BAR MANIA MINEIRA   

                                            ENDERECO  \
0  RUA DA BAHIA, 1144, CENTRO, BELO HORIZONTE, MG...   
1  RUA DO OURO, 835, SERRA, BELO HORIZONTE, MG, B...   
2  RUA PIAUI, 1884, SAVASSI, BELO HORIZONTE, MG, ...   
3  RUA MONTE ALEGRE, 160, SERRA, BELO HORIZONTE, ...   
4  RUA PARACATU, 1099, SANTO AGOSTINHO, BELO HORI...   

                                              IMAGEM                 PETISCO  \
0  https://s2-g1.glbimg.com/TBJlO2gBA-eoaA_0XK6kM...   Panela da Diversidade   
1  https://s2-g1.glbimg.com/4nvx0TTUnKRrUtN5CL_bO...  Lingua da vovó Virinha   
2  https://s2-g1.glbimg.com/jzIpvnVuWznTE1n1R0us6...          Canelada Suína   
3  https://s2-g1.glbimg.com/XMgVOHr-0_Y9lUNhH3Fr4...                Tipracas   
4  https://cdb-static-files.s3.amazonaws.com/wp-c...       EXPLOSÃO DE SABOR   

                                           DESCRICAO  
0  Acem cozido com ervas finas acompanhado com ba...  
1  Língua de boi assada sobre creme de batata e m...  
2  Panturrilha suína, mandioca cozida com queijo ...  
3  Pelotas de rabada com angu de agrião e "tiprac...  
4            Bolinho de mandioca recheado com queijo

### Arrumando manulamente os dados do Comida di Buteco que não teve as coordenas achadas pela API

In [ ]:
import pandas as pd
import math
import numpy as np

def encontrar_bares_sem_coords(df):
    def parse_coord(x):
        if isinstance(x, tuple):
            return x
        if isinstance(x, str):
            x = x.strip().replace("(", "").replace(")", "")
            parts = x.split(",")
            if len(parts) != 2:
                return (np.nan, np.nan)
            try:
                return (float(parts[0]), float(parts[1]))
            except:
                return (np.nan, np.nan)
        return (np.nan, np.nan)

    def is_nan_tuple(t):
        return (
            isinstance(t, tuple) and len(t) == 2 and
            math.isnan(t[0]) and math.isnan(t[1])
        )

    df["ID_CDB"] = df["ID_CDB"].astype(str)

    df["COORDS_parsed"] = df["COORDS"].apply(parse_coord)
    df["COORD_GEO_parsed"] = df["COORD_GEO"].apply(parse_coord)

    # Filtros
    filtro_cdb = df["ID_CDB"] != "000"
    filtro_coords_nan = df["COORDS_parsed"].apply(is_nan_tuple)
    filtro_coordgeo_nan = df["COORD_GEO_parsed"].apply(is_nan_tuple)

    resultado = df.loc[
        filtro_cdb & filtro_coords_nan & filtro_coordgeo_nan,
        ["NOME", "ENDERECO_COMPLETO"]
    ].reset_index(drop=True)

    df.drop(columns=["COORDS_parsed", "COORD_GEO_parsed"], inplace=True)

    return resultado


In [ ]:
df = pd.read_csv("complete_bar_data.csv")
resultado = encontrar_bares_sem_coords(df)
print(resultado)


                 NOME                                  ENDERECO_COMPLETO
0        BUTECO'S BAR  RUA ERNESTO BRAGA, 2, JARDIM ATLANTICO, BELO H...
1   DECK BOI NA BRASA  RUA DESEMBARGADOR JOSÉ SATYRO, 302, MANACAS, B...
2   COMPANHIA DO DINO  RUA JOAQUIM CLEMENTE, 682, FLORAMAR, BELO HORI...


BUTECO'S BAR: Rua Ernesto Braga 02 Jardim Atlântico, Belo Horizonte -> -19.8478398,-43.9870588
DECK BOI NA BRASA: Rua Desembargador José Satyro, 302 - Castelo, Belo Horizonte -> -19.8876361!4d-43.9964471
COMPANHIA DO DINO: Avenida Joaquim Clemente, 682 - Floramar, Belo Horizonte -> -19.8372406!4d-43.9332652

Pesquisei no Google Maps e adicionei manualmente ao complete_bar_data.csv, assim garantimos que pelo menos COORDS ou COORD_GEO tem alguma coordenada, logo sempre conseguimos printar os bares